### PR TITLE
Adding tic-tac-toe and car-races from cb

### PR DESCRIPTION
**Pull request by MedovTimur**

### CODE REVIEW

File: contracts/car-races/assets/accounts_pr_keys.txt
Comment (shamilsan) [inline contracts/car-races/assets/accounts_pr_keys.txt:1]: Do we need there files in the repo?

File: contracts/car-races/car-1/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#![no_std]
+use gstd::{collections::BTreeMap, msg, prelude::*, ActorId};
+
+#[derive(Encode, Decode, TypeInfo)]
+#[codec(crate = gstd::codec)]
+#[scale_info(crate = gstd::scale_info)]
+pub enum CarAction {
+    YourTurn(BTreeMap<ActorId, Car>),
+}
+#[derive(Encode, Decode, TypeInfo, Clone)]
```
Comment (shamilsan) [inline contracts/car-races/car-1/src/lib.rs:10]: Too dense.
```suggestion

#[derive(Encode, Decode, TypeInfo, Clone)]
```

File: contracts/car-races/io/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    ///
+    /// The unit tuple is used as we don't process any signals in this contract.
+    type Signal = ();
+
+    type State = InOut<StateQuery, StateReply>;
+}
+
+#[derive(Encode, Decode, TypeInfo)]
+#[codec(crate = gstd::codec)]
+#[scale_info(crate = gstd::scale_info)]
```
Comment (shamilsan) [inline contracts/car-races/io/src/lib.rs:86]: Too many types here, and these attributes look dirty. Maybe it's worth to import them from `codec` and `scale_info`.

File: contracts/car-races/io/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+    GasReserved,
+    StrategyAdded,
+    PlayersMove,
+}
+
+impl Game {
+    pub fn buy_acceleration(&mut self) {
+        let car_id = self.get_current_car_id();
+        let car = self.cars.get_mut(&car_id).expect("Get Car: Can't be None");
+        car.speed += DEFAULT_SPEED;
```
Comment (shamilsan) [inline contracts/car-races/io/src/lib.rs:242]: We prefer checked or saturated adding to avoid overflow.

File: contracts/car-races/io/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+        let car_id = self.get_current_car_id();
+
+        let shelled_car_id = self.find_car_to_shell(&car_id);
+        self.cars.entry(shelled_car_id).and_modify(|car| {
+            let new_speed = car.speed - DEFAULT_SPEED;
+            if new_speed < DEFAULT_SPEED {
+                car.speed = DEFAULT_SPEED;
+            } else {
+                car.speed = new_speed;
+            }
```
Comment (shamilsan) [inline contracts/car-races/io/src/lib.rs:256]: We could use `max` here:
```suggestion
            car.speed = new_speed.max(DEFAULT_SPEED);
```

File: contracts/car-races/io/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+        car.speed += DEFAULT_SPEED;
+        car.car_actions.push(RoundAction::Accelerated);
+    }
+
+    pub fn buy_shell(&mut self) {
+        let car_id = self.get_current_car_id();
+
+        let shelled_car_id = self.find_car_to_shell(&car_id);
+        self.cars.entry(shelled_car_id).and_modify(|car| {
+            let new_speed = car.speed - DEFAULT_SPEED;
```
Comment (shamilsan) [inline contracts/car-races/io/src/lib.rs:251]: Isn't underflow possible here if `car.speed` is less than `DEFAULT_SPEED`?

File: contracts/car-races/io/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+            car.car_actions.push(RoundAction::SlowedDown);
+        });
+    }
+
+    pub fn get_current_car_id(&self) -> ActorId {
+        self.car_ids[self.current_turn as usize]
+    }
+
+    pub fn update_positions(&mut self, config: &Config) {
+        let mut winners = Vec::new();
```
Comment (shamilsan) [inline contracts/car-races/io/src/lib.rs:266]: Could we use capacity equal to 3 here?
```suggestion
        let mut winners = Vec::with_capacity(3);
```

File: contracts/car-races/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+            exec::program_id(),
+            "Only program can send this message"
+        );
+
+        let game = self.get_game(account);
+
+        if game.state == GameState::Finished {
+            let result = game.result.clone();
+            let cars = game.cars.clone();
+            let car_ids = game.car_ids.clone();
```
Comment (shamilsan) [inline contracts/car-races/src/lib.rs:119]: Why clone if they are just borrowed later?

File: contracts/car-races/tests/races_node_test.rs [hunk 8]
```diff
@@ -0,0 +1,208 @@
+// use car_races_io::*;
+// use futures::executor::block_on;
+// use gclient::{EventProcessor, GearApi, Result, WSAddress};
```
Comment (shamilsan) [inline contracts/car-races/tests/races_node_test.rs:3]: Dead code? It is better to delete it or uncomment.

**Review by shamilsan — APPROVED** [review]
Tests must have. Text files with keys are not to be added in my opinion.

**Review by YauheniDraichykau — APPROVED** [review]

**Pull request merged**